orchestrator/orchestrator/src/main_loop.rs

```diff
@@ -134,82 +134,72 @@ pub async fn eth_oracle_main_loop(
     let mut grpc_client = grpc_client;
 
     loop {
-        let loop_start = Instant::now();
-
-        let latest_eth_block = web3.eth_block_number().await;
-        let latest_cosmos_block = contact.get_chain_status().await;
-        match (latest_eth_block, latest_cosmos_block) {
-            (Ok(latest_eth_block), Ok(ChainStatus::Moving { block_height })) => {
-                metrics::set_cosmos_block_height(block_height.clone());
-                metrics::set_ethereum_block_height(latest_eth_block.clone());
-                trace!(
-                    "Latest Eth block {} Latest Cosmos block {}",
-                    latest_eth_block,
-                    block_height,
-                );
-            }
-            (Ok(_latest_eth_block), Ok(ChainStatus::Syncing)) => {
-                warn!("Cosmos node syncing, Eth oracle paused");
-                delay_for(DELAY).await;
-                continue;
```

Comment:
Maybe we should consider using `tokio::time::interval` ([docs](https://docs.rs/tokio/1.13.0/tokio/time/fn.interval.html)) in place of sleep. It will pause and only continue once a tick elapses every amount of time specified. What differentiates it from `sleep` according to the docs is that it measures the time passed since the last tick, rather than beginning its countdown when the line is hit. Therefore, instead of using `join!`, we would have something like

```rust
loop {
    // create new interval
    let interval = tokio::time::interval(Duration::from_secs(ETH_ORACLE_LOOP_SPEED));
    // first tick is immediate
    interval.tick();

    let latest_eth_block = web3.eth_block_number().await;
    let latest_cosmos_block = contact.get_chain_status().await;
    match (latest_eth_block, latest_cosmos_block) {
        (Ok(latest_eth_block), Ok(ChainStatus::Moving { block_height })) => {
            metrics::set_cosmos_block_height(block_height.clone());
            metrics::set_ethereum_block_height(latest_eth_block.clone());
            trace!(
                "Latest Eth block {} Latest Cosmos block {}",
                latest_eth_block,
                block_height,
            );
        // we can still use continue since we aren't using join!
        continue;
        // etc...
    }

    // This will pause until ETH_ORACLE_LOOP_SPEED has elapsed *since the 
    // start of the loop at the first tick*
    interval.tick().await;
}
```

We can also configure how it should behave if the second tick() line is hit after the specified time has already elapsed. By default it burst-executes all missed ticks I think.

Reply:
The problem here isn't `sleep`, rather it's `join`.

Reply:
I know. I am saying that `interval` is a better solution than `join` because it allows us to `continue` the loop while still making sure the loop only runs every `ETH_ORACLE_LOOP_SPEED` seconds if the `match` is a successful case. 

The auditor said "using Instant::now() leads to non-exact timeouts and loop_speed because the checks are always if the elapsed time is greater or lesser than a fixed Duration", but this is solved by `interval`.

Reply:
In my opinion, I'll rather use `timeout` and `sleep`. Let's discuss this on Monday